src/query/mod.rs

```diff
@@ -0,0 +1,46 @@
+mod query_parse;
+mod query_types;
+mod query_eval;
+mod query_format;
+
+extern crate pest;
+
+use crate::index::get_packet_index;
+use std::fmt;
+use crate::query::query_parse::{parse_query, Rule};
+use crate::query::query_eval::eval_query;
+use crate::query::query_format::format_query_result;
+
+pub fn run_query(root: &str, query: String) -> Result<String, QueryError> {
+    let index = match get_packet_index(root) {
+        Ok(index) => index,
+        Err(e) => {
+            return Err(QueryError::EvalError(format!(
+                "Could not build outpack index from root at {}: {:?}",
+                root, e
+            )))
+        }
+    };
+    let parsed = parse_query(&query)?;
+    let result = eval_query(index, parsed);
+    format_query_result(result)
+}
+
+#[derive(Debug, Clone)]
+// Results with QueryError are at least as large as the QueryError variant. The compiler
+// will need to reserve that much memory every time it is used. We want to keep this as
+// small as possible so Box the large error body to force it onto the heap.
+// See https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err
+pub enum QueryError {
+    ParseError(Box<pest::error::Error<Rule>>),
```

Comment:
Add a constructor that will box the error to avoid having all the Box's throughout here (if that is possible)

Reply:
I think no straightforward enum variant ctor. I could write another function to build this from the `Rule` but then that is passing the unboxed variable into a function, but then I think (?) Rust will potentially copy the potentially large `pest::error::Error` when moving it into the function. After reading through this weeks Rust chapter it looks like this is the idiomatic way to box a variant on an enum.